src/fsharp/FSharp.Core/prim-types.fs

```diff
@@ -6497,6 +6497,95 @@ namespace Microsoft.FSharp.Core
                   interface IEnumerable with 
                       member x.GetEnumerator() = (gen() :> IEnumerator) }
 
+            [<NoEquality; NoComparison>]
+            type VariableStepIntegralRangeState<'a> = {
+                mutable Started  : bool
+                mutable Complete : bool
+                mutable Current  : 'a
+            }
+            let inline variableStepIntegralRange n step m =
+                if step = LanguagePrimitives.GenericZero
+                then invalidArg "step" (SR.GetString(SR.stepCannotBeZero));
+                else
+                    let variableStepRangeEnumerator () =
+                        let state = {
+                            Started  = false
+                            Complete = false
+                            Current  = Unchecked.defaultof<'a>
+                        }
+
+                        { new System.Collections.Generic.IEnumerator<'a> with
+                            member __.Dispose () = ()
+
+                            member __.Current =
+                                // according to IEnumerator<int>.Current documentation, the result of of Current
+                                // is undefined prior to the first call of MoveNext and post called to MoveNext
+                                // that return false (see https://msdn.microsoft.com/en-us/library/58e146b7%28v=vs.110%29.aspx)
+                                // so we should be able to just return value here, and we could get rid of the 
+                                // complete variable which would be faster
+                                if not state.Started then notStarted ()
+                                elif state.Complete  then alreadyFinished ()
+                                else state.Current
+
+                            member this.Current =
+                                box this.Current
```

Comment:
why `box`? `this.Current` it' not of generic `IEnumerator<'a>` so `abstract Current : 'a with get`?